lib/kino/smart_cell/sql.ex

```diff
@@ -0,0 +1,200 @@
+            %{variable: conn_attrs["variable"], type: conn_attrs["type"]}
+          end,
+        result_variable: attrs["result_variable"] || "result"
+      )
+
+    {:ok, ctx, editor: [attribute: "query", language: "sql"]}
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    payload = %{
+      connections: ctx.assigns.connections,
+      connection: ctx.assigns.connection,
+      result_variable: ctx.assigns.result_variable
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_event("update_connection", variable, ctx) do
+    connection = Enum.find(ctx.assigns.connections, &(&1.variable == variable))
+    ctx = assign(ctx, connection: connection)
+    broadcast_event(ctx, "update_connection", connection.variable)
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_result_variable", variable, ctx) do
+    ctx =
+      if Kino.Utils.Code.valid_variable_name?(variable) do
+        broadcast_event(ctx, "update_result_variable", variable)
+        assign(ctx, result_variable: variable)
+      else
+        broadcast_event(ctx, "update_result_variable", ctx.assigns.result_variable)
+        ctx
+      end
+
+    {:noreply, ctx}
+  end
+
+  @impl true
+  def scan_binding(pid, binding, _env) do
+    connections =
+      for {key, value} <- binding,
+          is_atom(key),
+          type = connection_type(value),
+          do: %{variable: Atom.to_string(key), type: type}
+
+    send(pid, {:connections, connections})
+  end
+
+  @impl true
+  def handle_info({:connections, connections}, ctx) do
+    connections = Enum.sort_by(connections, & &1.variable)
+
+    connection =
+      case {connections, ctx.assigns.connection} do
+        {[connection | _], nil} -> connection
+        {_connections, connection} -> connection
+      end
+
+    broadcast_event(ctx, "connections", %{
+      "connections" => connections,
+      "connection" => connection
+    })
+
+    {:noreply, assign(ctx, connections: connections, connection: connection)}
+  end
+
+  @compile {:no_warn_undefined, {DBConnection, :connection_module, 1}}
+
+  defp connection_type(connection) when is_pid(connection) do
+    with true <- Code.ensure_loaded?(DBConnection),
+         {:ok, module} <- DBConnection.connection_module(connection) do
+      case Atom.to_string(module) do
+        "Elixir.Postgrex" <> _ -> "postgres"
+        "Elixir.MyXQL" <> _ -> "mysql"
+        _ -> nil
+      end
+    else
+      _ -> nil
+    end
+  end
+
+  defp connection_type(_connection), do: nil
+
+  @impl true
+  def to_attrs(ctx) do
+    %{
+      "connection" =>
+        if connection = ctx.assigns.connection do
+          %{"variable" => connection.variable, "type" => connection.type}
+        end,
+      "result_variable" => ctx.assigns.result_variable
+    }
+  end
+
+  @impl true
+  def to_source(attrs) do
+    attrs
+    |> to_quoted()
+    |> Macro.to_string()
+    # TODO: remove reformatting on Elixir v1.14 (before v1.13.1
```

Comment:
Macro.to_string is using the same foundation as `Code.format_string` in Elixir v1.13, so you can use the direct APIs instead: https://github.com/elixir-lang/elixir/blob/main/lib/elixir/lib/macro.ex#L971-L974